modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+    }
+    else {
+      if debugStrings then
+        chpl_debug_string_print("  lhs remote: "+lhs.locale.id:c_string);
+      on lhs do helpMe(lhs, rhs);
+    }
+
+    if debugStrings then
+      chpl_debug_string_print("leaving proc =()");
   }
 
-  //
-  // casts from c_string to integer types
-  //
-  pragma "insert line file info"
-  extern proc c_string_to_int8_t  (x:c_string) : int(8); 
-  pragma "insert line file info"
-  extern proc c_string_to_int16_t (x:c_string) : int(16);
-  pragma "insert line file info"
-  extern proc c_string_to_int32_t (x:c_string) : int(32);
-  pragma "insert line file info"
-  extern proc c_string_to_int64_t (x:c_string) : int(64);
-  pragma "insert line file info"
-  extern proc c_string_to_uint8_t (x:c_string) : uint(8); 
-  pragma "insert line file info"
-  extern proc c_string_to_uint16_t(x:c_string) : uint(16);
-  pragma "insert line file info"
-  extern proc c_string_to_uint32_t(x:c_string) : uint(32);
-  pragma "insert line file info"
-  extern proc c_string_to_uint64_t(x:c_string) : uint(64);
-
-  inline proc _cast(type t, x:c_string) where t == int(8)
-    return c_string_to_int8_t(x);
-  inline proc _cast(type t, x:c_string) where t == int(16)
-    return c_string_to_int16_t(x);
-  inline proc _cast(type t, x:c_string) where t == int(32)
-    return c_string_to_int32_t(x);
-  inline proc _cast(type t, x:c_string) where t == int(64)
-    return c_string_to_int64_t(x);
-  inline proc _cast(type t, x:c_string) where t == uint(8)
-    return c_string_to_uint8_t(x);
-  inline proc _cast(type t, x:c_string) where t == uint(16)
-    return c_string_to_uint16_t(x);
-  inline proc _cast(type t, x:c_string) where t == uint(32)
-    return c_string_to_uint32_t(x);
-  inline proc _cast(type t, x:c_string) where t == uint(64)
-    return c_string_to_uint64_t(x);
+  proc =(ref lhs: string, rhs_c: c_string) {
+    if debugStrings then
+      chpl_debug_string_print("in proc =() c_string");
 
-  //
-  // casts from c_string to real/imag types
-  //
-  pragma "insert line file info"
-  extern proc c_string_to_real32(x:c_string) : real(32);
-  pragma "insert line file info"
-  extern proc c_string_to_real64(x:c_string) : real(64);
-  pragma "insert line file info"
-  extern proc c_string_to_imag32(x:c_string) : imag(32);
-  pragma "insert line file info"
-  extern proc c_string_to_imag64(x:c_string) : imag(64);
-  
-  inline proc _cast(type t, x:c_string) where t == real(32)
-    return c_string_to_real32(x);
-  inline proc _cast(type t, x:c_string) where t == real(64)
-    return c_string_to_real64(x);
-  inline proc _cast(type t, x:c_string) where t == imag(32)
-    return c_string_to_imag32(x);
-  inline proc _cast(type t, x:c_string) where t == imag(64)
-    return c_string_to_imag64(x);
+    // Make this some sort of local check once we have local types/vars
+    if (rhs_c.locale.id != chpl_nodeID) || (lhs.locale.id != chpl_nodeID) then
+      halt("Cannot assign a remote c_string to a string.");
```

Comment:
Why not? Just make LHS have owned set and be a copy of RHS. You can copy a c_string.